From 441497e7ffbf58d46a2316acf41c09477d5bd06a Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Wed, 24 Jun 2026 15:17:13 +0800
Subject: [PATCH] mtk: hostapd: AP MLD: clear STA's neg_ttlm in
 hostapd_teardown_neg_ttlm()

Moving the clearing logic inside hostapd_teardown_neg_ttlm() prevents
issues when a caller fails to do so.

fixup: mtk: hostapd: add Neg-TTLM support for AP MLD

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/ieee802_11_eht.c | 8 ++------
 1 file changed, 2 insertions(+), 6 deletions(-)

diff --git a/src/ap/ieee802_11_eht.c b/src/ap/ieee802_11_eht.c
index 426df1d8e..970189723 100644
--- a/src/ap/ieee802_11_eht.c
+++ b/src/ap/ieee802_11_eht.c
@@ -1967,7 +1967,6 @@ void hostapd_neg_ttlm_resp_tx_status(struct hostapd_data *hapd,
 	if (ret) {
 		ieee802_11_send_neg_ttlm_teardown(hapd, assoc_sta->addr);
 		hostapd_teardown_neg_ttlm(assoc_hapd, assoc_sta);
-		os_memset(neg_ttlm, 0, sizeof(*neg_ttlm));
 	}
 }
 
@@ -3186,7 +3185,7 @@ fail:
 	if (ret) {
 		/* For synchronization */
 		ieee802_11_send_neg_ttlm_teardown(hapd, addr);
-		os_memset(&sta->neg_ttlm, 0, sizeof(sta->neg_ttlm));
+		hostapd_teardown_neg_ttlm(assoc_hapd, sta);
 	}
 }
 
@@ -3218,7 +3217,6 @@ static void ieee802_11_rx_neg_ttlm_resp(struct hostapd_data *hapd, const u8 *add
 		if (ret) {
 			ieee802_11_send_neg_ttlm_teardown(hapd, sta->addr);
 			hostapd_teardown_neg_ttlm(assoc_hapd, sta);
-			os_memset(neg_ttlm, 0, sizeof(*neg_ttlm));
 		}
 	} else {
 		wpa_printf(MSG_ERROR, "TTLM: STA rejected TTLM request");
@@ -3242,8 +3240,6 @@ static void ieee802_11_rx_neg_ttlm_teardown(struct hostapd_data *hapd, const u8
 	if (!sta)
 		return;
 
-	os_memset(&sta->neg_ttlm, 0, sizeof(sta->neg_ttlm));
-
 	hostapd_teardown_neg_ttlm(assoc_hapd, sta);
 	return;
 }
@@ -3633,7 +3629,7 @@ void hostapd_teardown_neg_ttlm(struct hostapd_data *hapd, struct sta_info *sta)
 {
 	const u8 *addr = sta->addr;
 
-	sta->neg_ttlm.valid = false;
+	os_memset(&sta->neg_ttlm, 0, sizeof(sta->neg_ttlm));
 
 	if (hostapd_is_attlm_active(hapd) && !hapd->mld->new_attlm.switch_time) {
 		struct attlm_settings *attlm = &hapd->mld->new_attlm;
-- 
2.45.2

